src/nl/hannahsten/texifyidea/editor/folding/LatexFootnoteFoldingBuilder.kt

```diff
@@ -0,0 +1,46 @@
+package nl.hannahsten.texifyidea.editor.folding
+
+import com.intellij.lang.ASTNode
+import com.intellij.lang.folding.FoldingBuilderEx
+import com.intellij.lang.folding.FoldingDescriptor
+import com.intellij.openapi.editor.Document
+import com.intellij.openapi.project.DumbAware
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiElement
+import com.intellij.refactoring.suggested.endOffset
+import com.intellij.refactoring.suggested.startOffset
+import nl.hannahsten.texifyidea.psi.LatexCommands
+import nl.hannahsten.texifyidea.psi.LatexRequiredParam
+import nl.hannahsten.texifyidea.util.childrenOfType
+import nl.hannahsten.texifyidea.util.firstChildOfType
+
+/**
+ * Adds folding regions for LaTeX environments.
+ *
+ * Enables folding of `\footnote{}`.
+ *
+ * @author jojo2357
+ */
+class LatexFootnoteFoldingBuilder : FoldingBuilderEx(), DumbAware {
+
+    override fun isCollapsedByDefault(node: ASTNode) = true
+
+    override fun getPlaceholderText(node: ASTNode): String {
+        val parsedText = node.text.substring(1).trim()
+        return if (parsedText.length > 8) parsedText.substring(0, 8) + "..." else parsedText.substring(0, parsedText.length - 1)
+    }
+
+    override fun buildFoldRegions(root: PsiElement, document: Document, quick: Boolean): Array<FoldingDescriptor> {
+        val descriptors = ArrayList<FoldingDescriptor>()
+        val environments =
+            root.childrenOfType(LatexCommands::class).filter { it.commandToken.text == "\\footnote" }.mapNotNull {
+                it.firstChildOfType(LatexRequiredParam::class)
+            }
+
+        for (environment in environments) {
+            descriptors.add(FoldingDescriptor(environment.originalElement, TextRange(environment.startOffset + 1, environment.endOffset - 1)))
```

Comment:
Take care of boundaries and edge cases.

```
java.lang.AssertionError: (202,202), text: {}, language = Language: Latex
	at com.intellij.lang.folding.FoldingDescriptor.<init>(FoldingDescriptor.java:178)
	at com.intellij.lang.folding.FoldingDescriptor.<init>(FoldingDescriptor.java:102)
	at com.intellij.lang.folding.FoldingDescriptor.<init>(FoldingDescriptor.java:83)
	at com.intellij.lang.folding.FoldingDescriptor.<init>(FoldingDescriptor.java:68)
	at com.intellij.lang.folding.FoldingDescriptor.<init>(FoldingDescriptor.java:55)
	at nl.hannahsten.texifyidea.editor.folding.LatexFootnoteFoldingBuilder.buildFoldRegions(LatexFootnoteFoldingBuilder.kt:41)
```

Reply:
I don't know how you did that, but i added a check for latex

Reply:
Well, if `startOffset + 1 >= endOffset - 1`, for example, you cannot fold nothing

Reply:
Oh that breaks it? I was just going to let it be. If you can't tell I mostly just recycled the existing folders

Reply:
Nevermind you were not referring to the update I just pushed.